x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/AnnPQQuery.java

```diff
@@ -0,0 +1,157 @@
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                final VectorDocValues values = (VectorDocValues) context.reader().getBinaryDocValues(fieldType.name());
+                return new ANNPQScorer(this, values);
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                return null;
+            }
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return true;
+            }
+        };
+    }
+
+    private class ANNPQScorer extends Scorer {
+        private final VectorDocValues values;
+        private final float[][] cqDistances; // squared distances between queryVector and all product centroids
+
+        private ANNPQScorer(Weight w, VectorDocValues values) throws IOException {
+            super(w);
+            this.values = values;
+
+            // precalculate squared distances between queryVector and all product centroids
+            // we use euclidean distance for this
+            // ((c1-q1)^2 + (c2-q2)^2 ...) can be converted to (||c||^2 + ||q||^2 - 2*c*q)
+            // where ||c||^2 -- squared magnitude of product centroid
+            // where ||q||^2 -- squared magnitude of query vector
+            float[][][] pCentroids = values.pcentroids();
+            float[][] pcSquaredMagnitudes = values.pCentroidsSquaredMagnitudes();
+            int pcentroidsCount = values.pcentroidsCount();
+
+            this.cqDistances = new float[pqCount][pcentroidsCount];
+            for (int pq = 0; pq < pqCount; pq++) {
+                for (int c = 0; c < pcentroidsCount; c++) {
+                    float dotProduct = 0;
+                    for (int dim = 0; dim < pdims; dim++) {
+                        dotProduct += queryVector[pq * pdims + dim] * pCentroids[pq][c][dim];
+                    }
+                    cqDistances[pq][c] = pcSquaredMagnitudes[pq][c] + qSquareMagnitudes[pq] - 2 * dotProduct;
+                }
+            }
+        }
+
+        @Override
+        public final int docID() {
+            return values.docID();
+        }
+
+        @Override
+        public float score() throws IOException {
+            BytesRef valueBR = values.docCentroids();
+            int offset = valueBR.offset;
+            float score = 0;
+            for (int i = 0; i < pqCount; i++) {
+                int docCentroid = valueBR.bytes[offset++] & 0xFF; // centroids are stored as unsigned bytes
+                score += cqDistances[i][docCentroid];
+            }
+            score = 1/score; // as we want to get closest vectors fist, score is inversely proportional to distance
+            return score;
+        }
+
+        @Override
+        public final DocIdSetIterator iterator() {
+            return values;
+        }
+
+        @Override
+        public float getMaxScore(int upTo) {
+            return Float.MAX_VALUE;
```

Comment:
We know the max score so it could speed up boolean queries to return the correct one here ?

Reply:
It could also be interesting to avoid some computations if the current max score (setCompetitiveScore) is not reachable without some centroids. That could work like conjunctions where we check that the sum of the clauses can reach the minimum competitive score except that here each centroid have a static score.